### DISCUSSION TITLE
fork-point and squash

### DISCUSSION THREAD

**Discussion by levinotik** [VirtusLab/git-machete#1067] [Q&A] [answered]
The help for some of the commands is sometimes a bit difficult to grasp so I'm having some trouble understanding `fork-point` and how it interacts with `squash`. 

In a young repo, I wanted to squash all of my commits. `git machete squash` outputs:

```shell
Fork point not found for branch master; use git machete fork-point master --override-to...
```

(not sure if that's being ellipsized due to my shell or if that's the full output)

Maybe I'm missing something, but`fork-point --help` didn't seem to provide enough info for me to know what exactly I need to do to be able to `squash`. I'm sure I'm just missing a basic concept here. 

So, in this case — newish repo, one `master` branch, and 20 commits, how can I simply squash everything down to the first commit? 

Is the main concept here that since squash will squash the unique history of the branch, if there is no other branch machete doesn't know which commit to consider the divergence point in order to select the correct set of commits?

Appreciate any guidance and explanation!

**Comment by PawelLipski** [comment, marked as answer]
👋🏻 

Hey there... is `master` a root branch (i.e. a branch that does **not** have any parent in .git/machete)? It's likely that it's the same UX problem as in https://github.com/VirtusLab/git-machete/issues/930 🤔 I'll fix that soon 🔧 

Anyway, you can try `git machete squash --fork-point=HEAD~20` for now!

**Reply by levinotik** [reply]
Thanks for the help. That's correct `.git/machete` is just `master`:

```
❯ cat .git/machete
master
```

I think I did basically what you had suggested, but with `git machete fork-point --override-to=30bae04` where `30bae04` was either the first or second commit (can't remember). 

Thank you for the reply! Still trying to wrap my head around machete.

**Reply by PawelLipski** [reply]
Also...

>  I wanted to squash all of my commits

but then

> git machete fork-point --override-to=30bae04 where 30bae04 was either the first or second commit 

I suppose you haven't managed to squash **all** commits eventually, right? only **all but one** or all but two, right? 🤔 I've realized there's indeed no support for squashing all commits (as it's impossible to specify the virtual "zeroth" commit as fork point via commit hash)

**Reply by PawelLipski** [reply]
> Is the main concept here that since squash will squash the unique history of the branch, if there is no other branch machete doesn't know which commit to consider the divergence point in order to select the correct set of commits?

And yes, indeed, that's the concept :)
